flake.nix

```diff
@@ -87,18 +87,37 @@
     flake-compat.flake = false;
   };
 
-  outputs = { self, nixpkgs, flake-utils, emacs-overlay, ... }@inputs:
-    let inherit (flake-utils.lib) eachDefaultSystem eachSystem;
+  outputs = { self, nixpkgs, flake-utils, ... }@inputs:
+    let inherit (flake-utils.lib) eachDefaultSystem mkApp;
     in eachDefaultSystem (system:
       let pkgs = import nixpkgs { inherit system; };
       in {
-        devShell = pkgs.mkShell {
+        apps = {
+          default = self.outputs.apps.${system}.doom-emacs-example;
+          doom-emacs-example = mkApp {
+            drv = self.outputs.packages.${system}.doom-emacs-example;
+            exePath = "/bin/emacs";
+          };
+        };
+
+        devShells.default = pkgs.mkShell {
           buildInputs =
             [ (pkgs.python3.withPackages (ps: with ps; [ PyGithub ])) ];
         };
-        package = { dependencyOverrides ? { }, ... }@args:
-          pkgs.callPackage self
-          (args // { dependencyOverrides = (inputs // dependencyOverrides); });
+
+        package = { ... }@args:
+          pkgs.lib.warn ''
+            Deprecated, will be removed after NixOS 23.05 release.
+            Please use `packages.${system}.default` instead!
+          ''
+          (pkgs.callPackage self args);
```

Comment:
But those two are not equivalent? `default` is just the example? 

How are you supposed to pass in your own arguments using a flake output this way? `override`?

Reply:
Yeah, using `override`. #266 has instructions on how to use it properly.

Reply:
I improved the message to include that you can use `override` here instead.

Reply:
Scary how this means we expose `test/doom.d` as a part of our API. Wouldn't it be weird if you "installed" NDE and then the default configuration changed and it "migrated" to your "installation"? I think the previous behavior was better as it kind of guided you into setting the config right away. `override` is not nearly as obvious.

Reply:
> Wouldn't it be weird if you "installed" NDE and then the default configuration changed and it "migrated" to your "installation"?

I don't think so, the documentation in #267 can be upgraded to make it clear that this is only an example.

Also, I don't think the `test/doom.d` is a really good example neither: for example, it doesn't enable many modules, and more importantly, it doesn't enable any language support. So almost any kinda of file you open with it (with the exception of Elisp) has zero syntax highlighting.

So not sure which kinda of users would wanted to use this config as-is.

> I think the previous behavior was better as it kind of guided you into setting the config right away. override is not nearly as obvious.

The problem with the previous behavior is that it is non-standard: you shouldn't expose functions as `package`/`packages` output, it **must** be derivations. This non-standard usage causes all sort of small breakages (`nix flake show`/`nix flake checks` for example), and it even cause confusion for users that know what they should expect (this is one of the things @MonAaraj complained in Matrix).

Reply:
@ckiee I can add an info/warn when the user is using the `test/doom.d`, guiding them to use `override` and setting their own config instead. WDYT?

Reply:
This would also trigger in `app` output, so it would also solve the other issue

Reply:
> @ckiee I can add an info/warn when the user is using the `test/doom.d`, guiding them to use `override` and setting their own config instead. WDYT?

I did a few tests, this doesn't look like a good idea. Adding this kinda of message ends up polluting many outputs like `nix flake show` and `nix flake checks`.

Seems like `nix` needs a proper `builtins.println` instead of just having `builtins.trace` (but on the other side, maybe this is a bad idea too :P).

Reply:
Going over everything again I think the change shouldn't be made here; instead we should make the `./test/doom.d` config start up with a [`y-or-n-p`](https://ftp.gnu.org/old-gnu/Manuals/elisp-manual-21-2.8/html_node/elisp_281.html) asking if you want to learn more about Doom Emacs pointing to some docs in the nix store. This will be better than before where the API guided you into discovering there are other arguments you can pass in.

Reply:
> Going over everything again I think the change shouldn't be made here; instead we should make the `./test/doom.d` config start up with a [`y-or-n-p`](https://ftp.gnu.org/old-gnu/Manuals/elisp-manual-21-2.8/html_node/elisp_281.html) asking if you want to learn more about Doom Emacs pointing to some docs in the nix store. This will be better than before where the API guided you into discovering there are other arguments you can pass in.

I am all for it, but doing all of this in this PR would be too much.

And I would much prefer to have the Flake refactor now than later, because not only it makes things more difficult for newcomers, it also breaks nix tooling itself.

Reply:
Also, with both this and #266, I think we will be already be in a much better position for beginners. This one brings a more familiar Flake output, while #266 should really bring us better documentation.

Reply:
> I am all for it, but doing all of this in this PR would be too much.

Created #276.

> And I would much prefer to have the Flake refactor now than later, because not only it makes things more difficult for newcomers, it also breaks nix tooling itself.

Aw, come on, that nix tooling argument is a total strawman (;

Most of the PR would be good to merge now but I think the API change should wait until #266 is done with both of our reviews since the commits are kind of clumped badly for this. In hindsight, "flake.nix cleanup" and a "API/onboarding refactor" PRs would have served us better. Up to you if you want to wait or move the API change into #266/a new PR, though I don't mind playing with git if you're exhausted. (: